### PR TITLE
Move key ranges via postgres_fdw

### PR DESCRIPTION
**Pull request by EinKrebs**

### CODE REVIEW

File: coordinator/provider/coordinator.go [hunk 1]
```diff
@@ -812,63 +834,87 @@ func (qc *qdbCoordinator) Move(ctx context.Context, req *kr.MoveKeyRange) error
...
-	moveId, err := qc.RecordKeyRangeMove(ctx,
-		&qdb.MoveKeyRange{
-			MoveId:     uuid.New().String(),
-			ShardId:    req.ShardId,
-			KeyRangeID: req.Krid,
-			Status:     qdb.MoveKeyRangePlanned,
-		})
-
-	/* NOOP if key range move was already recorded */
+	move, err := qc.GetKeyRangeMove(ctx, req.Krid)
```
Comment (Denchick) [inline coordinator/provider/coordinator.go:837]: Lets add a comment here (or below) what `nil` move means

File: pkg/datatransfers/data_transfers.go [hunk 2]
```diff
@@ -81,243 +77,177 @@ MoveKeys performs physical key-range move from one datashard to another.
...
+	}
 	if shards == nil {
 		err := LoadConfig(config.CoordinatorConfig().ShardDataCfg)
 		if err != nil {
 			spqrlog.Zero.Error().Err(err).Msg("error loading config")
 		}
 	}
 
-	from, err := pgx.Connect(ctx, createConnString(fromId))
+	from, err := pgx.Connect(ctx, createConnString(fromId)) // nolint: staticcheck
```
Comment (Denchick) [inline pkg/datatransfers/data_transfers.go:106]: Why not to fix it?

File: qdb/memqdb.go [hunk 3]
```diff
@@ -427,7 +428,7 @@ func (q *MemQDB) GetTransferTx(ctx context.Context, key string) (*DataTransferTr
 
 	ans, ok := q.Transactions[key]
 	if !ok {
-		return nil, spqrerror.Newf(spqrerror.SPQR_TRANSFER_ERROR, "no tx with key %s", key)
+		return nil, nil
```
Comment (Denchick) [inline qdb/memqdb.go:431]: the wat

File: pkg/util/uitl.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import (
+	"fmt"
+	"github.com/pg-sharding/spqr/pkg/models/distributions"
+	"github.com/pg-sharding/spqr/pkg/models/kr"
+	"strings"
+)
+
+// GetKRCondition returns SQL condition for elements of distributed relation between two key ranges
+// TODO support multidimensional key ranges
+func GetKRCondition(ds *distributions.Distribution, rel *distributions.DistributedRelation, kRange *kr.KeyRange, upperBound kr.KeyRangeBound, prefix string) string {
```
Comment (Denchick) [inline pkg/util/uitl.go:12]: I don't understand two things here

1. What this method do and what `KRConditions` is.
2. Why it is in `pkg/util/uitl.go`? It looks like not an util method.

btw in filename there is a typo

**Review by Denchick — CHANGES_REQUESTED** [review]

File: pkg/datatransfers/data_transfers.go (hunk 2 above)
Comment (EinKrebs) [reply pkg/datatransfers/data_transfers.go:105]: It's a false positive

File: coordinator/provider/coordinator.go [hunk 5]
```diff
@@ -342,16 +342,15 @@ func (qc *qdbCoordinator) RunCoordinator(ctx context.Context, initialRouter bool
...
-			Msg("faild to list key ranges")
+			Msg("failed to list key ranges")
 	}
 
 	for _, r := range ranges {
 		tx, err := qc.db.GetTransferTx(context.TODO(), r.KeyRangeID)
 		if err != nil {
 			continue
 		}
-		if tx.ToStatus == qdb.Commited && tx.FromStatus != qdb.Commited {
```
[the remark this answers is not in the document]
Comment (EinKrebs) [reply coordinator/provider/coordinator.go:353]: Logic of movement has changed, tx status is checked in (*qbdCoordinator).Move method

File: pkg/datatransfers/data_transfers_test.go [hunk 6]
```diff
@@ -1,173 +1,12 @@
...
-	"github.com/golang/mock/gomock"
-	"github.com/jackc/pgx/v5"
-	"github.com/jackc/pgx/v5/pgconn"
-	mock "github.com/pg-sharding/spqr/pkg/mock/pgx"
-	"github.com/pg-sharding/spqr/qdb"
 	"github.com/stretchr/testify/assert"
 )
 
-// commitTransactions tests
-func TestCommitPositive(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (EinKrebs) [reply pkg/datatransfers/data_transfers_test.go:18]: There's no transactions to commit now

File: pkg/util/uitl.go (hunk 4 above)
Comment (EinKrebs) [reply pkg/util/uitl.go]: What this method does is described in comment above. Also I moved it to kr package

File: pkg/datatransfers/data_transfers.go (hunk 2 above)
Comment (EinKrebs) [reply pkg/datatransfers/data_transfers.go]: Lint magically stoped failing here, so I removed the nolint

File: qdb/memqdb.go (hunk 3 above)
Comment (EinKrebs) [reply qdb/memqdb.go:431]: There's no need of an error if transaction does not exist

**Review by Denchick — CHANGES_REQUESTED** [review]

File: qdb/etcdqdb.go [hunk 7]
```diff
@@ -448,32 +457,36 @@ func (q *EtcdQDB) RecordTransferTx(ctx context.Context, key string, info *DataTr
 
 // TODO : unit tests
 func (q *EtcdQDB) GetTransferTx(ctx context.Context, key string) (*DataTransferTransaction, error) {
-	resp, err := q.cli.Get(ctx, key, clientv3.WithPrefix())
+	spqrlog.Zero.Debug().
+		Str("key", key).
+		Msg("etcdqdb: get data transfer tx")
+
+	// TODO move to separate namespace
```
[the remark this answers is not in the document]
Comment (EinKrebs) [reply qdb/etcdqdb.go:464]: This TODO is obsolete, gonna remove it

File: qdb/etcdqdb.go [hunk 8]
```diff
@@ -56,6 +56,7 @@ const (
 	shardsNamespace          = "/shards/"
 	relationMappingNamespace = "/relation_mappings/"
 	taskGroupPath            = "/move_task_group"
+	transactionNamespace     = "/transfer_txs/"
```
[the remark this answers is not in the document]
Comment (EinKrebs) [reply qdb/etcdqdb.go:59]: What is now called transfer transaction is an act of transfering actual data between shards. Key range move also includes locking-modifying-unlocking the key range. Also, I think `transaction_transfers` is not that good of a name. Maybe stuff like `data_transfers` or `key_transfers` would be better?

File: pkg/models/kr/keyrange.go [hunk 9]
```diff
@@ -104,3 +107,40 @@ func (kr *KeyRange) ToProto() *proto.KeyRangeInfo {
 		DistributionId: kr.Distribution,
 	}
 }
+
+// GetKRCondition returns SQL condition for elements of distributed relation between two key ranges
+// TODO support multidimensional key ranges
+func GetKRCondition(ds *distributions.Distribution, rel *distributions.DistributedRelation, kRange *KeyRange, upperBound KeyRangeBound, prefix string) string {
```
Comment (Denchick) [inline pkg/models/kr/keyrange.go:113]: This is a great method to write unit tests on.

**Review by Denchick — CHANGES_REQUESTED** [review]

**Review by Denchick — APPROVED** [review]

**Pull request merged**